apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/RetryHandler.cs

```diff
@@ -37,46 +36,47 @@ internal sealed class RetryHandler : IHttpUnsuccessfulResponseHandler
         internal const string AttemptCountHeaderPart = "gccl-attempt-count";
 
-        private static readonly int[] s_retriableErrorCodes =
-        {
-                408, // Request timeout
-                429, // Too many requests
-                500, // Internal server error
-                502, // Bad gateway
-                503, // Service unavailable
-                504 // Gateway timeout
-        };
-        private static readonly RetryHandler s_instance = new RetryHandler();
+        private readonly RetryOptions _retryOptions;
 
-        private RetryHandler() { }
+        internal RetryHandler(RetryOptions retryOptions) => _retryOptions = retryOptions;
 
+        // TODO: Will remove it once the implementation across all APIs is complete
         internal static void MarkAsRetriable<TResponse>(StorageBaseServiceRequest<TResponse> request)
         {
+            RetryHandler retryHandler = new RetryHandler(RetryOptions.IdempotentRetryOptions);
+
             // Note: we can't use ModifyRequest, as the x-goog-api-client header is added later by ConfigurableMessageHandler.
             // Additionally, that's only called once, and we may want to record the attempt number as well.
             request.AddExecuteInterceptor(InvocationIdInterceptor.Instance);
-            request.AddUnsuccessfulResponseHandler(s_instance);
+            request.AddUnsuccessfulResponseHandler(retryHandler);
         }
 
-        // This function is designed to support asynchrony in case we need to examine the response content, but for now we only need the status code
-        internal static Task<bool> IsRetriableResponse(HttpResponseMessage response) => 
-            Task.FromResult(s_retriableErrorCodes.Contains(((int)response.StatusCode)));
+        internal static void MarkAsRetriable<TResponse>(StorageBaseServiceRequest<TResponse> request, RetryOptions options)
+        {
+            //In case null retry options or Never is passed, retry will not happen in case of failure
+            if (options == null || options.Predicate == RetryPredicate.Never || options == RetryOptions.Never)
+            {
+                return;
+            }
+            RetryHandler retryHandler = new RetryHandler(options);
+
+            // Note: we can't use ModifyRequest, as the x-goog-api-client header is added later by ConfigurableMessageHandler.
+            // Additionally, that's only called once, and we may want to record the attempt number as well.
+            request.AddExecuteInterceptor(InvocationIdInterceptor.Instance);
+            request.AddUnsuccessfulResponseHandler(retryHandler);
+        }
 
         public async Task<bool> HandleResponseAsync(HandleUnsuccessfulResponseArgs args)
         {
-            var retry = args.SupportsRetry && await IsRetriableResponse(args.Response).ConfigureAwait(false);
+            var retry = (_retryOptions != null) && args.SupportsRetry && _retryOptions.Predicate.ShouldRetry((int) args.Response.StatusCode);
```

Comment:
We don't need the nullity test now, as we'll never create an instance of RetryHandler where `_retryOptions` is null.